Plutarch/Builtin.hs

```diff
@@ -264,16 +297,16 @@ pforgetData = pupcast
 {- | Like 'pforgetData', except it works for complex types.
  Equivalent to 'pupcastF'.
 -}
-pforgetData' :: forall a (p :: PType -> PType) s. Proxy p -> Term s (p (PAsData a)) -> Term s (p PData)
-pforgetData' _ = punsafeCoerce
+pforgetData' :: forall a (p :: PType -> PType) s. PCovariant p => Proxy p -> Term s (p (PAsData a)) -> Term s (p PData)
+pforgetData' = pupcastF
 
 -- | Inverse of 'pforgetData''.
-prememberData :: forall (p :: PType -> PType) s. Proxy p -> Term s (p PData) -> Term s (p (PAsData PData))
-prememberData Proxy = punsafeCoerce
+prememberData :: forall (p :: PType -> PType) s. PVariant p => Proxy p -> Term s (p PData) -> Term s (p (PAsData PData))
+prememberData Proxy = let _ = witness (Proxy @(PVariant p)) in punsafeCoerce
 
 -- | Like 'prememberData' but generalised.
-prememberData' :: forall a (p :: PType -> PType) s. PSubtype PData a => Proxy p -> Term s (p a) -> Term s (p (PAsData a))
-prememberData' Proxy = punsafeCoerce
+prememberData' :: forall a (p :: PType -> PType) s. (PSubtype PData a, PVariant p) => Proxy p -> Term s (p a) -> Term s (p (PAsData a))
+prememberData' Proxy = let _ = witness (Proxy @(PSubtype PData a, PVariant p)) in punsafeCoerce
```

Comment:
do `PIsDataRepr` instances have a `PSubtype` relationship with `PData`?

Reply:
That doesn't exist anymore.

Reply:
ah, I see. Well, is there such a subtype relationship between `PlutusType` derived via `PlutusTypeData` then?

Reply:
Yep